src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala

```diff
@@ -1,37 +1,223 @@
+    storage.setQuantity(assetAtTransaction, quantity)
+  }
+
+  def assetRollbackTo(assetId: AssetId, height: Int): Unit = {
+    val asset = Base58.encode(assetId)
+
+    val heights = storage.getHeights(asset)
+    val heightsToRemove = heights.filter(h => h > height)
+    storage.setHeight(asset, heights -- heightsToRemove)
+
+    val transactionsToRemove: Seq[String] = heightsToRemove.foldLeft(Seq.empty[String]) { (result, h) =>
+      result ++ storage.getTransactions(s"$asset@$h")
+    }
+
+    val keysToRemove = transactionsToRemove.map(t => s"$asset@$t")
+
+    keysToRemove.foreach { key =>
+      storage.removeKey(key)
+    }
+  }
+
+  def isReissuable(assetId: AssetId): Boolean = {
+    val asset = Base58.encode(assetId)
+    val heights = storage.getHeights(asset)
+
+    heights.lastOption match {
+      case Some(lastHeight) =>
+        val transactions = storage.getTransactions(s"$asset@$lastHeight")
+        if (transactions.nonEmpty) {
+          val transaction = transactions.last
+          storage.isReissuable(s"$asset@$transaction")
+        } else false
+      case None => false
+    }
+  }
+
+  def getAssetName(assetId: AssetId): String = {
+    storage.getTransaction(assetId).flatMap {
+      case tx: IssueTransaction => Some(tx.asInstanceOf[IssueTransaction])
+      case _ => None
+    }.map(tx => new String(tx.name, Charsets.UTF_8)).getOrElse("Unknown")
+  }
+
+  def applyExchangeTransaction(blockTs: Long)(tx: Transaction): Unit = tx match {
+    case om: ExchangeTransaction =>
+      def isSaveNeeded(order: Order): Boolean = {
+        order.expiration >= blockTs
+      }
+
+      def putOrder(order: Order) = {
+        if (isSaveNeeded(order)) {
+          val orderDay = calcStartDay(order.expiration)
+          storage.putSavedDays(orderDay)
+          val orderIdStr = Base58.encode(order.id)
+          val omIdStr = Base58.encode(om.id)
+          val prev = storage.getOrderMatchTxByDay(orderDay, orderIdStr).getOrElse(Array.empty[String])
+          if (!prev.contains(omIdStr)) {
+            storage.putOrderMatchTxByDay(orderDay, orderIdStr, prev :+ omIdStr)
+          }
+        }
+      }
+
+      def removeObsoleteDays(timestamp: Long): Unit = {
+        val ts = calcStartDay(timestamp)
+        val daysToRemove: List[Long] = storage.savedDaysKeys.filter(t => t < ts)
+        if (daysToRemove.nonEmpty) {
+          synchronized {
+            storage.removeOrderMatchDays(daysToRemove)
+          }
+        }
+      }
+
+      putOrder(om.buyOrder)
+      putOrder(om.sellOrder)
+      removeObsoleteDays(blockTs)
+
+    case _ =>
+  }
+
+  def findPrevOrderMatchTxs(order: Order): Set[ExchangeTransaction] = {
+
+    def parseTxSeq(a: Array[String]): Set[ExchangeTransaction] = {
+      a.toSet.flatMap { s: String => Base58.decode(s).toOption }.flatMap { id =>
+        storage.getTransactionBytes(id).flatMap(b => ExchangeTransaction.parseBytes(b).toOption)
+      }
```

Comment:
This could easily be replaced with for-comprehension, which is much more readable.